packages/core/src/event-store/postgres/store.ts

```diff
@@ -373,46 +376,132 @@ export class PostgresEventStore implements EventStore {
-        ([startBlock, endBlock]) => ({
-          factoryId,
-          startBlock: BigInt(startBlock),
-          endBlock: BigInt(endBlock),
-        })
-      );
+          const existingIntervals = await tx
+            .deleteFrom("factoryLogFilterIntervals")
+            .where("factoryId", "=", factoryId)
+            .returningAll()
+            .execute();
 
-      if (mergedIntervalRows.length > 0) {
-        await tx
-          .insertInto("factoryLogFilterIntervals")
-          .values(mergedIntervalRows)
-          .execute();
-      }
+          const mergedIntervals = intervalUnion(
+            existingIntervals.map((i) => [
+              Number(i.startBlock),
+              Number(i.endBlock),
+            ])
+          );
+
+          const mergedIntervalRows = mergedIntervals.map(
+            ([startBlock, endBlock]) => ({
+              factoryId,
+              startBlock: BigInt(startBlock),
+              endBlock: BigInt(endBlock),
+            })
+          );
+
+          if (mergedIntervalRows.length > 0) {
+            await tx
+              .insertInto("factoryLogFilterIntervals")
+              .values(mergedIntervalRows)
+              .execute();
+          }
+
+          return mergedIntervals;
+        });
+      })
+    );
 
-      return mergedIntervals;
+    const fragmentsWithIdx = fragments.map((f, idx) => ({ idx, ...f }));
+
+    const intervals = await this.db
+      .with(
+        "factoryFilterFragments(fragmentIndex, fragmentAddress, fragmentEventSelector, fragmentChildAddressLocation, fragmentTopic0, fragmentTopic1, fragmentTopic2, fragmentTopic3)",
+        () =>
+          sql`( values ${sql.join(
+            fragmentsWithIdx.map(
+              (f) =>
+                sql`( ${sql.val(f.idx)}, ${sql.val(f.address)}, ${sql.val(
+                  f.eventSelector
+                )}, ${sql.val(f.childAddressLocation)}, ${sql.val(
+                  f.topic0
+                )}, ${sql.val(f.topic1)}, ${sql.val(f.topic2)}, ${sql.val(
+                  f.topic3
+                )} )`
+            )
+          )} )`
+      )
+      .selectFrom("factoryLogFilterIntervals")
+      .leftJoin("factories", "factoryId", "factories.id")
+      .innerJoin("factoryFilterFragments", (join) => {
+        let baseJoin = join.on(({ or, cmpr }) =>
+          or([
+            cmpr("address", "is", null),
+            cmpr("fragmentAddress", "=", sql.ref("address")),
+          ])
+        );
+        baseJoin = join.on(({ or, cmpr }) =>
+          or([
+            cmpr("eventSelector", "is", null),
+            cmpr("fragmentEventSelector", "=", sql.ref("eventSelector")),
+          ])
+        );
+        baseJoin = join.on(({ or, cmpr }) =>
+          or([
+            cmpr("childAddressLocation", "is", null),
+            cmpr(
+              "fragmentChildAddressLocation",
+              "=",
+              sql.ref("childAddressLocation")
+            ),
+          ])
+        );
```

Comment:
Unlike with log filters address + topics, I think these values will always be defined. So, we can probably replace this with:
```ts
let baseJoin = join.on(({ or, cmpr }) =>
  and([
    cmpr("fragmentAddress", "=", sql.ref("address")),
    cmpr("fragmentEventSelector", "=", sql.ref("eventSelector")),
    cmpr("fragmentChildAddressLocation", "=", sql.ref("childAddressLocation"))
  ])
);
```
The topics below are still optional, so that block looks good.

Reply:
Nice catch, pretty sure the second line of your snippet is supposed to be `or`.

Reply:
Never mind, I think the first line is supposed to have `and` as a function param instead of `or`